[12.0] [MIG] l10n_ch_payment_slip  + l10n_ch_base_bank

i-vyshnevska opened the pull request:
pick latest changes from v 11.0 and update to v12

yvaucher committed:
[MIG] l10n_ch_payment_slip: Migration to 11.0

Update headers
Port code to python3
Update README
Fix tests with correct report render calls
Adapt code to report module split

yvaucher committed:
Rename BVR and BVR/ESR to ISR in english and make it the base term

grindtildeath committed:
Improve tests and assign bvr_address_format without sudo

grindtildeath committed:
Improve tests and assign bvr_address_format without sudo

simahawk committed:
l10n_ch_payment_slip: fix `_get_address_lines`

`payment_slip._get_address_lines` computes the address of the partner.

This is done in an onchange context whereas the changes where not isolated
and they were propagated to the real partner
via all the onchanges related to it.

We spotted the issue on the mail compose wizard
whereas the chain of onchanges was wiping all the default values.

With this change we:

* isolate the partner w/ a fake record
* additionally we cache the result per each partner
  since the function is called several times

grindtildeath committed:
Allow payment slip to be saved in and read from attachment

grindtildeath committed:
Delete payment slip attachment when setting invoice back to draft

grindtildeath committed:
Add test case on attachment management

TDu committed:
Fix some docstrings, typos and english language

TDu committed:
Fix test for python3

TDu committed:
Fix non qweb report not supported

TDu committed:
Remove v11 import

It has been moved to a own module l10n_ch_import_isr_v11

TDu committed:
Improve ISR file name

TDu committed:
Add python dependency

TDu committed:
Simplify js code

TDu committed:
Fix batch print ISR

TDu committed:
Move parameters to setting view

grindtildeath committed:
Fix v11 for attachment feature

Commit:
[11.0][IMP] Remove useless button on invoice

Remove the button "Print ISR" on invoices to avoid confusion with Odoo base
implementation and rename the one from this module for user consistency

yvaucher committed:
Migration of attachment feature to v11

grindtildeath committed:
Fix dependency to remove std print isr button

Commit:
Fix right error on ISR print

grindtildeath committed:
Finish migration to 11.0

Restore report_type to reportlab-pdf

Improvement to batch wizard, error readonly

grindtildeath committed:
Fix merges methods

grindtildeath committed:
Rework save as attachment feature using _post_pdf

grindtildeath committed:
Drop unneeded imports

grindtildeath committed:
Finally fix the test considering odoo weird behaviour

oca-travis committed:
[UPD] Update l10n_ch_payment_slip.pot

chrispi-ch committed:
Translated using Weblate (German)

Currently translated at 100.0% (25 of 25 strings)

Translation: l10n-switzerland-11.0/l10n-switzerland-11.0-l10n_ch_base_bank
Translate-URL: https://translation.odoo-community.org/projects/l10n-switzerland-11-0/l10n-switzerland-11-0-l10n_ch_base_bank/de/

chrispi-ch committed:
Translated using Weblate (German)

Currently translated at 93.3% (70 of 75 strings)

Translation: l10n-switzerland-11.0/l10n-switzerland-11.0-l10n_ch_payment_slip
Translate-URL: https://translation.odoo-community.org/projects/l10n-switzerland-11-0/l10n-switzerland-11-0-l10n_ch_payment_slip/de/

MaxyMoos committed:
[FIX] res.partner.bank: retrieve partner_name from context on creation (#428)

asaunier committed:
[MIG] l10n_ch_base_bank: Migration to odoo 12

Commit:
[MIG][12.0] improvements to v12

i-vyshnevska commented:
fails due to  https://github.com/OCA/l10n-switzerland/pull/465
and https://github.com/odoo/odoo/pull/31293

Tonow-c2c commented on l10n_ch_payment_slip/report/ir_action.py:
```diff
@@ -1,46 +1,172 @@
+                                  res_ids=res_ids), 'pdf'
+        return pdf_content, 'pdf'
+
+    # TODO consider https://github.com/OCA/reporting-engine/issues/241
```
Time to do this TODO ? [https://github.com/OCA/reporting-engine/issues/241](https://github.com/OCA/reporting-engine/issues/241)
or remove if it's no more relevant

yvaucher replied:
This payment slip is about to die, the new format using QR code will be used in few months.

I think we can drop the TODO.

Martronic-SA committed:
small and not samll

It's just syntax but more readable

Commit:
l10n_ch_base_bank: update tests and invoice model

Changet in test new() to create and write() to run test properly
update tetst cases due to errors, refactor behavior of updating
acc_number and ccp

Due to updates in  account_payment_partner 95174ee
changet from direct access to method partner_banks_to_show

Commit:
[IMP] use common method from bank-payment for bank searching

oca-travis committed:
[UPD] Update l10n_ch_base_bank.pot

Commit:
[IMP] cleanup pylint

simahawk commented on l10n_ch_bank/__init__.py:
```diff
@@ -1,12 +1,8 @@
+from .hooks import import_csv_data
 
 
 def post_init(cr, registry):
```
`post_init` should stay in `hooks.py` and simply be imported here

simahawk commented on l10n_ch_base_bank/__manifest__.py:
```diff
@@ -1,21 +1,24 @@
+ 'version': '12.0.1.0.0',
  'author': "Camptocamp,Odoo Community Association (OCA)",
  'category': 'Localization',
  'website': 'http://www.camptocamp.com',
```
https://github.com/OCA/l10n-switzerland

simahawk commented on l10n_ch_base_bank/__manifest__.py:
```diff
@@ -1,21 +1,24 @@
      'views/bank.xml',
      'views/invoice.xml',
  ],
  'demo': [],
```
remove empty/default keys

simahawk commented on l10n_ch_base_bank/models/bank.py:
```diff
@@ -152,8 +152,8 @@ def name_search(self, name, args=None, operator='ilike', limit=80):
             ids = self.search(
                 args,
                 limit=limit
-                ).ids
```
@yvaucher can't we build a big domain search for all the cols in OR and do only one search?

yvaucher replied:
I find useless to search an empty value on all fields.
But it can be improved by constructing only the domain in the `if...else` and pass the domain to the same search.

``` python
if name:
    domain = [...] + args # using name
else:
    domain = args
ids = self._search(domain, limit=limit, ...)
```

We might even use `_search` to get directly the `ids` without using `browse` implied by `search`

simahawk commented on l10n_ch_base_bank/models/bank.py:
```diff
@@ -236,29 +272,53 @@ def _check_postal_num(self):
+            part_name = self.env['res.partner'].browse(partner_id)[0].name
+        self.acc_number = self._compute_name_ccp(part_name, self.ccp)
+
+    @api.multi
```
seems more `api.model`

simahawk commented on l10n_ch_base_bank/models/bank.py:
```diff
@@ -303,58 +365,65 @@ def onchange_acc_number_set_swiss_bank(self):
+        """If ccp changes and it's a postal bank update acc_number to ccp
+        we don't want make acc_number as computed to have possibility set it
+        manually and also avoid to shadow other logic on acc_number if exist
+       """
```
indentation seems weird here

simahawk commented on l10n_ch_base_bank/models/bank.py:
```diff
@@ -303,58 +365,65 @@ def onchange_acc_number_set_swiss_bank(self):
-                else:
-                    self.acc_number = self._get_acc_name()
+    def onchange_ccp_set_acc_number(self):
+        """If ccp changes and it's a postal bank update acc_number to ccp
```
this is more a comment than a docstring

simahawk commented on l10n_ch_base_bank/models/bank.py:
```diff
@@ -303,58 +365,65 @@ def onchange_acc_number_set_swiss_bank(self):
+        if ccp and self.is_swiss_postal_num(ccp) and not self.bank_id.id:
+            bank = (self.env['res.bank'].search([('ccp', '=', ccp)], limit=1)
+                    or
+                    self.env['res.bank'].search([('bic', '=', 'POFICHBEXXX')],
```
this hardcoded value comes from?

i-vyshnevska replied:
that is bic for switzerland post, I move it to global var at least here

simahawk commented on l10n_ch_base_bank/models/bank.py:
```diff
@@ -303,58 +365,65 @@ def onchange_acc_number_set_swiss_bank(self):
 
     @api.onchange('partner_id')
     def onchange_partner_set_acc_number(self):
+        """
```
comment

simahawk commented on l10n_ch_base_bank/models/invoice.py:
```diff
@@ -121,15 +117,40 @@ def _check_isr(self):
+        """
+        res = super().write(vals)
+        for rec in self:
+            if not rec.partner_bank_id and rec.type == 'out_invoice':
```
any chance to avoid multiple write?

i-vyshnevska replied:
we want use here partner_banks_to_show() which comes from bank-payment repo and works only with self

simahawk commented on l10n_ch_base_bank/models/invoice.py:
```diff
@@ -121,15 +117,40 @@ def _check_isr(self):
+
     @api.model
     def create(self, vals):
         """We override create in order to have customer invoices
```
```py
# make sure to have customer invoices generated by the commercial flow
# since on partner change it doesn't happen automatically
```

simahawk commented on l10n_ch_base_bank/readme/CONTRIBUTORS.rst:
```diff
@@ -0,0 +1,7 @@
+* Guewen Baconnier <guewen.baconnier@camptocamp.com>
+* Paul Catinean <paulcatinean@gmail.com>
+* Yannick Vaucher <yannick.vaucher@camptocamp.com>
+* Akim Juillerat <akim.juillerat@camptocamp.com>
```
after this effort you can put yourself here no? :)

simahawk commented on l10n_ch_base_bank/readme/USAGE.rst:
```diff
@@ -0,0 +1,8 @@
+Account type will be discovered authomatically.
```
*automatically

simahawk commented on l10n_ch_base_bank/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,7 @@
+This addons will add different bank types required by specific swiss electronic
```
*Swiss

simahawk commented on l10n_ch_base_bank/readme/DESCRIPTION.rst:
```diff
@@ -0,0 +1,7 @@
+This addons will add different bank types required by specific swiss electronic
+payment like DTA and ISR. It allows to manage both Post and Bank systems.
+
+It'll perform some validation when entring bank account number or ESR number
```
*entering or filling

simahawk commented on l10n_ch_base_bank/tests/test_bank.py:
```diff
@@ -12,8 +12,27 @@
+    def setUp(self):
+        super(TestBank, self).setUp()
+        self.partner = self.env.ref('base.res_partner_12')
+        self.bank = self.env['res.bank'].create({
```
do we need to do this for _every_ test? If not pls use SavepointCase and set tracking_disable globally

simahawk commented on l10n_ch_base_bank/tests/test_bank.py:
```diff
@@ -158,46 +174,43 @@ def test_set_ccp(self):
         self.assertEqual(bank_acc.bank_id, self.post_bank)
 
     def test_set_ccp_bank(self):
+        # we crate bank account
```
*create

simahawk commented on l10n_ch_base_bank/tests/test_bank.py:
```diff
@@ -282,68 +303,79 @@ def test_name_search(self):
+        # everything cleanup
+        bank_acc.ccp = ''
+        bank_acc._update_acc_name()
+        self.assertEqual(bank_acc.acc_number, '')
```
in all these tests I see onchanges called manually: in v12 we have a new common class to ease form behaviors testing https://www.odoo.com/documentation/12.0/reference/testing.html#odoo.tests.common.Form
might be worth to give it a try.
If you don't go for it at least drop a comment for future improvement ;)

i-vyshnevska replied:
I saw it, but not in this PR, and what I saw in v12 they have improved testing for onchange by default

simahawk commented on l10n_ch_base_bank/tests/test_create_invoice.py:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2015 Yannick Vaucher (Camptocamp SA)
+# Copyright 2012-2019 Camptocamp
```
same story regarding SavepointCase + tracking_disable
+ pls move setup of the test class at the top (it makes no sense to have the preparation of the suite at the bottom)

simahawk commented on l10n_ch_base_bank/tests/test_search_invoice.py:
```diff
@@ -7,23 +7,31 @@
 class TestSearchInvoice(common.TransactionCase):
 
     def setUp(self):
-        super(TestSearchInvoice, self).setUp()
```
same

simahawk commented on l10n_ch_payment_slip/__manifest__.py:
```diff
@@ -1,32 +1,36 @@
+ 'version': '12.0.1.0.0',
  'author': "Camptocamp,Odoo Community Association (OCA)",
  'category': 'Localization',
  'website': 'http://www.camptocamp.com',
```
https://github.com/OCA/l10n-switzerland

simahawk commented on l10n_ch_payment_slip/__manifest__.py:
```diff
@@ -1,32 +1,36 @@
      "report/report_declaration.xml",
      "security/ir.model.access.csv"
  ],
  'demo': [],
```
same about default/empty flags

simahawk commented on l10n_ch_payment_slip/controllers/web.py:
```diff
@@ -0,0 +1,34 @@
+class ReportController(report.ReportController):
+    @route()
+    def report_routes(self, reportname, docids=None, converter=None, **data):
+        if converter == "reportlab-pdf":
```
is reportlab still something in the modern era? :thinking:

simahawk commented on l10n_ch_payment_slip/models/invoice.py:
```diff
@@ -169,3 +171,24 @@ def _check_bvr_generatable(self):
+            report_payment_slip = False
+        if report_payment_slip and report_payment_slip.attachment:
+            for invoice in self:
+                with invoice.env.do_in_draft():
```
Q: why in draft mode? You are not changing anything, no?

i-vyshnevska replied:
it comes from https://github.com/OCA/l10n-switzerland/pull/411 I don't want improve it in this PR

simahawk replied:
perfectly fine :+1: but please leave comments in all the places where it would be nice to rethink/refactor in the future, thanks!

simahawk commented on l10n_ch_payment_slip/models/payment_slip.py:
```diff
@@ -27,7 +25,7 @@ class PaymentSlipSettings(object):
 
     def __init__(self, report_name, **kwargs):
-        for param, value in kwargs.iteritems():
+        for param, value in list(kwargs.items()):
```
no need to convert to list

simahawk commented on l10n_ch_payment_slip/report/ir_action.py:
```diff
@@ -1,46 +1,170 @@
+        return report
+
+    @api.multi
+    def render_reportlab_pdf(self, res_ids=None, data=None):
```
is this still used?

simahawk commented on l10n_ch_payment_slip/static/src/js/qwebactionmanager.js:
```diff
@@ -0,0 +1,37 @@
+        ir_actions_report: function (action, options) {
+            var report_url = '';
+
+            if (action.report_type !== 'reportlab-pdf') {
```
same for reportlab story

i-vyshnevska replied:
@yvaucher  can you pls add some context about reportlab here?

yvaucher replied:
@i-vyshnevska reportlab was chosen to generate the pdf for precision issue with printers.

We have multiple parameters to align the payment slip and it's elements vertically and horizontally. Print-out might be altered by software configuration or printer setup if not printer hardware itself.

With Qweb it was way to hard to align correctly the payment slip and the Swiss post specs would make your payment slip rejected if you don't fix in a tiny error margin.

Nowadays, it may have improved and the payment slip as we know it will soon(-ish) be replaced (starting 30th of June 2020) by the QRCode version for which the readers will be more efficient.

i-vyshnevska replied:
@yvaucher I  added it's to comments for clarifying, can we pass this PR further?

simahawk commented on l10n_ch_payment_slip/views/report_xml_templates.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
rename this file to `assets.xml`

simahawk commented on l10n_ch_zip/__init__.py:
```diff
@@ -1,12 +1,8 @@
+from .hooks import import_csv_data
 
 
 def post_init(cr, registry):
```
same as per base module: move to hooks

simahawk commented:
@i-vyshnevska this PR is huge, it takes too long to go thru it. Can't we split the 2 modules in 2 PRs?

i-vyshnevska commented:
> @i-vyshnevska this PR is huge, it takes too long to go thru it. Can't we split the 2 modules in 2 PRs?

payment_slip was partially migrated and now failing, when push base_bank, I started them separately but  to keep green merge

yvaucher commented on l10n_ch_base_bank/models/bank.py:
```diff
@@ -186,6 +214,17 @@ class ResPartnerBank(models.Model, BankCommon):
+
+    @api.model
+    def _get_supported_account_types(self):
+        rslt = super(ResPartnerBank, self)._get_supported_account_types()
```
Just variable naming here.

I guess `rslt` is for result please prefer `res` or `result` which are more widely used in odoo and OCA.

yvaucher commented on l10n_ch_base_bank/models/invoice.py:
```diff
@@ -121,15 +117,39 @@ def _check_isr(self):
+            partner = self.env.user.company_id.partner_id
+            vals['partner_bank_id'] = self._get_bank_id(
+                partner, vals.get('journal_id'), vals.get('reference_type'),
+            )
```
In which case can't we rely on `partner_banks_to_show`?

Can't we let `partner_bank_id` undefined unless it is manually set by the user?

One reason we could want to copy the bank on invoice is for the user to see it.
Nevertheless, I would argue that the payment mode should already allow us to invoice partners by multiple means.
And that setting the bank on the invoice should be an exception and can be done manually.

I fear that we still are making something different than the OCA standard for other localisations. I would prefer to keep the partner_bank_id empty and only refer to `partner_banks_to_show`.

We probably want to find the payment mode instead.

i-vyshnevska replied:
this changes we made to satisfy this constrain https://github.com/OCA/l10n-switzerland/blob/37e940e6977ef0fc63b1fce526173a4cd5cd36f5/l10n_ch_base_bank/models/invoice.py#L72
partner_banks_to_show will work on already created obj

i-vyshnevska replied:
if we dropped constraint https://github.com/OCA/l10n-switzerland/blob/37e940e6977ef0fc63b1fce526173a4cd5cd36f5/l10n_ch_base_bank/models/invoice.py#L86
we can rid of create overriding, than only one test is failed https://github.com/OCA/l10n-switzerland/blob/37e940e6977ef0fc63b1fce526173a4cd5cd36f5/l10n_ch_base_bank/tests/test_create_invoice.py#L81
@yvaucher  maybe that test and behavior is obsolete?

i-vyshnevska replied:
@yvaucher ping

yvaucher replied:
How I see it:

1. As a user, I create an invoice.
2. I set the partner. (at this point the payment mode can be set by the onchange [1] )
3. I set the reference type to ISR
4. Then the constraint check if bank account (either on invoice or in payment mode) is compatible with ISR type.
   if yes -> nothing
   else -> raise error

In case of error, the user will have to change manually the payment mode or the partner_bank_id. (The error message needs to be adapted accordingly to explain that payment mode or invoice bank)

About `_get_bank_id`, it could be added to:
- the partner onchange of OCA/bank-payment/account_payment_partner
- plus a reference_type onchange.
But only if the reference type is ISR and it should also look at the payment modes.
Thus, I propose to drop it for the current PR and if the needs come we will add it in another PR.


TODO:

- Thus we can drop the part where we copy the bank account on the invoice `write` and `create`. I prefer to let the module `account_payment_partner`'s onchange do the job.
- Change the constraint to check with `partner_banks_to_show`
- Change the constraint message to reflect the possible use of either payment mode or bank field on invoice.
- Adapt the test you mentioned by calling the partner onchange or by setting the partner_bank_id (the check itself is about the constraint which is still valid)
- Drop `_get_bank_id` which could come in a further PR.


Reasons:

1. IMO we should stick to OCA/bank-payment behavior as much as possible as you could have a multi-company setup and you don't want the swiss company to mess with other country companies.
2. Less code make me more happy as long as we make it simple but not simpler :)


[1]
https://github.com/OCA/bank-payment/blob/1a21770d76ec3af9fa918b8256628257884113e9/account_payment_partner/models/account_invoice.py#L47-L49

yvaucher reviewed (changes requested):
Only one remaining question which is about the `partner_bank_id` field on invoice.

I would like the opinion of @fclementic2c here.

i-vyshnevska commented:
@yvaucher suggested changes are made in last fixup commit, also there is improvement for test cases  by using Form class, so it's a little bit messy again , sorry

yvaucher commented on l10n_ch_base_bank/models/invoice.py:
```diff
@@ -121,15 +112,22 @@ def _check_isr(self):
+    def partner_banks_to_show(self):
+        """
+        Extend method from account_payment_partner to add specific
+        logic for switzerland bank payments if base method give false result
```
I would reword it like this, as `false` could be interpreted as `wrong`

```suggestion
        logic for switzerland bank payments if base method does not give a result
```

yvaucher commented on l10n_ch_base_bank/models/invoice.py:
```diff
@@ -121,15 +112,22 @@ def _check_isr(self):
+        res = super().partner_banks_to_show()
+        if not res:
+            if self.journal_id:
+                return self.env['account.journal'].browse(self.journal_id.id)\
```
No need to rebrowse the object you already have.

```suggestion
                return self.journal_id.bank_account_id
```

yvaucher reviewed (changes requested):
Thanks for the changes.

Just spotted on comment and one line of code to improve.
